### PR TITLE
Start ci scans with a single request

### PR DESCRIPTION
**Pull request by jbergler**
We currently have 3 endpoints that the CLI must hit to start a scan.
While this only marginally increases the chance of failure if any one endpoint is intermittently unavailable, it makes it very hard to use the HTTP level metrics for scan endpoints to understand the success rate for starting a scan.

This switches the CLI to use a new endpoint that replaces all three requests previously required to start a scan and fetch configuration. It also uses strong types for the message format with the CLI.

### CODE REVIEW

File: cli/src/semgrep/app/scans.py [hunk 1]
```diff
@@ -48,17 +50,17 @@ class ScanCompleteResult:
 
 
 class ScanHandler:
```
Comment (aryx) [inline cli/src/semgrep/app/scans.py:49]: It would be great to remove this intermediate class actually. I'd rather build an actual object
when we received the scan_config ATD record from the backend.

Comment (aryx) [reply cli/src/semgrep/app/scans.py:49]: Right now we init() very early, with no info, and are forced to setup many fields as Optional and None, because the actual data we get in fetch_and_init_scan_config(), so I'd rather make this fetcn_and_init the init function.

Comment (jbergler) [reply cli/src/semgrep/app/scans.py:49]: Yeah, I agree this is a good change to make - we should do this separately though, the size of this PR is already getting pretty large just from the test refactoring, and I worry that doing both in one change will be even harder to review.

File: cli/src/semgrep/app/session.py [hunk 2]
```diff
@@ -155,8 +155,8 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
             ),
         )
 
-        self.mount("https://", retry_adapter)
-        self.mount("http://", retry_adapter)
+        # self.mount("https://", retry_adapter)
```
Comment (jbergler) [inline cli/src/semgrep/app/session.py]: TODO: figure out how to make requests-mock and this compatible.

File: cli/tests/e2e/snapshots/test_ci/test_outputs/noautofix---vim/results.txt [hunk 3]
```diff
@@ -30,36 +22,12 @@ poetry.lock:2:0:E:supply-chain1:found a dependency
...
-    https://semgrep.dev/orgs/org_name/findings
-    https://semgrep.dev/orgs/org_name/supply-chain
-  Has findings for blocking rules so exiting with code 1
-  semgrep.dev is suggesting a non-zero exit code <MASKED>
+  Initializing scan ...Traceback (most recent call last):
+  File "/Users/jbergler/co/semgrep-cli/cli/src/semgrep/commands/ci.py", line 321, in ci
+    scan_handler.start_scan(project_meta, project_config)
+  File "/Users/jbergler/co/semgrep-cli/cli/src/semgrep/app/scans.py", line 204, in start_scan
+    raise Exception(
+Exception: Failed to create a scan with given token and deployment_id.Please make sure they have been set correctly.API server at https://semgrep.dev returned this response: {"error":"The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again."}
```
Comment (jbergler) [inline cli/tests/e2e/snapshots/test_ci/test_outputs/noautofix---vim/results.txt]: TODO: fix me

File: cli/tests/e2e/snapshots/test_ci/test_outputs/noautofix---json/results.txt [hunk 4]
```diff
@@ -465,36 +22,12 @@ SEMGREP_APP_TOKEN="fake_key" SEMGREP_USER_AGENT_APPEND="pytest" SEMGREP_SETTINGS
...
-    https://semgrep.dev/orgs/org_name/findings
-    https://semgrep.dev/orgs/org_name/supply-chain
-  Has findings for blocking rules so exiting with code 1
-  semgrep.dev is suggesting a non-zero exit code <MASKED>
+  Initializing scan ...Traceback (most recent call last):
+  File "/Users/jbergler/co/semgrep-cli/cli/src/semgrep/commands/ci.py", line 321, in ci
+    scan_handler.start_scan(project_meta, project_config)
+  File "/Users/jbergler/co/semgrep-cli/cli/src/semgrep/app/scans.py", line 204, in start_scan
+    raise Exception(
+Exception: Failed to create a scan with given token and deployment_id.Please make sure they have been set correctly.API server at https://semgrep.dev returned this response: {"error":"The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again."}
```
Comment (jbergler) [inline cli/tests/e2e/snapshots/test_ci/test_outputs/noautofix---json/results.txt]: fix me

File: cli/src/semgrep/app/scans.py [hunk 5]
```diff
@@ -48,166 +47,123 @@ class ScanCompleteResult:
...
             unique_id=out.Uuid(self.local_id),
             requested_products=[],
+            dry_run=dry_run,
         )
-        self.scan_id = None
-        self.ignore_patterns: List[str] = []
-        self._policy_names: List[str] = []
-        self._autofix = False
-        self._deepsemgrep = False
+        self.scan_response: Optional[out.ScanResponse] = None
```
Comment (aryx) [inline cli/src/semgrep/app/scans.py:58]: once we change ScanHandler to be built at ScanResponse time, this field will never be None and we can actually remove lots of the boilerplace accessor code below.

File: cli/tests/e2e/snapshots/test_ci/test_existing_supply_chain_finding/config/base_output.txt [hunk 6]
```diff
@@ -34,8 +34,7 @@ SEMGREP_APP_TOKEN="fake_key" SEMGREP_USER_AGENT_APPEND="pytest" SEMGREP_SETTINGS
   environment - running in environment git, triggering event is unknown
 
   CONNECTION
-  Reporting start of scan for deployment_name (scan_id=<MagicMock name='post()…
-  Fetching configuration from Semgrep Cloud Platform
+  Initializing scan (deployment=org_name, scan_id=35976)
```
Comment (aryx) [inline cli/tests/e2e/snapshots/test_ci/test_existing_supply_chain_finding/config/base_output.txt]: we probably don't want to hardcode a scan_id in the snapshot.

File: cli/tests/e2e/snapshots/test_ci/test_full_run/autofix-azure-pipelines-overwrite-autodetected-variables/complete.json [hunk 7]
```diff
@@ -18,5 +18,6 @@
     },
     "engine_requested": "OSS"
   },
-  "dependency_parser_errors": []
+  "dependency_parser_errors": [],
+  "task_id": "324625d6-d4c7-4e22-8e2b-ae9b3f33bf42"
```
Comment (aryx) [inline cli/tests/e2e/snapshots/test_ci/test_full_run/autofix-azure-pipelines-overwrite-autodetected-variables/complete.json]: is this intended? Can we use a more obviously fake task_id if this is intended?

File: cli/tests/e2e/snapshots/test_ci/test_nosem/autofix---disable-nosem/output.txt [hunk 8]
```diff
@@ -10,7 +10,7 @@ SEMGREP_APP_TOKEN="" SEMGREP_USER_AGENT_APPEND="pytest" SEMGREP_SETTINGS_FILE="<
 
 
 ┌───────────────────────────┐
-│ 12 Blocking Code Findings │
+│ 10 Blocking Code Findings │
```
Comment (aryx) [inline cli/tests/e2e/snapshots/test_ci/test_nosem/autofix---disable-nosem/output.txt:13]: Why the change in this snapshot? Seems unrelated to the PR.

File: cli/tests/e2e/test_ci.py [hunk 9]
```diff
@@ -358,48 +354,84 @@ def automocks(mocker):
...
-            deployment_name="org_name",
-            ignored_files=[],
-            policy_names=["audit", "comment", "block"],
-            rule_config=file_content,
-        ),
+
+
+@pytest.fixture
+def mocked_scan_id() -> int:
+    return 35976
```
Comment (aryx) [inline cli/tests/e2e/test_ci.py]: Can we use a more obviously fake scan_id here? like 12345 ?

File: src/osemgrep/cli_ci/Ci_subcommand.ml [hunk 10]
```diff
@@ -179,6 +179,7 @@ let scan_config_and_rules_from_deployment ~dry_run
       unique_id = Uuidm.v `V4;
       (* TODO: should look at conf.secrets, conf.sca, conf.code, etc. *)
       requested_products = [];
+      dry_run;
```
Comment (aryx) [inline src/osemgrep/cli_ci/Ci_subcommand.ml]: maybe add a big TODO comment at the top of this file saying this is using the old way to request a scan and we should port the new code in scans.py that request a scan in one single POST request to /ap/cli/scans/

**Review by aryx — APPROVED** [review]
Looks great!

File: cli/tests/e2e/snapshots/test_ci/test_existing_supply_chain_finding/config/base_output.txt (hunk 6 above)
Comment (jbergler) [reply cli/tests/e2e/snapshots/test_ci/test_existing_supply_chain_finding/config/base_output.txt]: It's hard coded in the test fixtures so this should be fine.

File: cli/tests/e2e/test_ci.py (hunk 9 above)
Comment (jbergler) [reply cli/tests/e2e/test_ci.py]: Yep, good call.

Comment (jbergler) [reply cli/tests/e2e/test_ci.py]: yep, done

File: cli/tests/e2e/snapshots/test_ci/test_nosem/autofix---disable-nosem/output.txt (hunk 8 above)
Comment (jbergler) [reply cli/tests/e2e/snapshots/test_ci/test_nosem/autofix---disable-nosem/output.txt:13]: This is a side effect of the tests now consistently using the same ignored finding match based ID's - I think it's worth letting this change just since it makes the tests a bit more consistent.

I confirmed that the disabled vs enabled cases still have the coverage we want.

File: cli/tests/e2e/snapshots/test_ci/test_full_run/autofix-azure-pipelines-overwrite-autodetected-variables/complete.json (hunk 7 above)
Comment (jbergler) [reply cli/tests/e2e/snapshots/test_ci/test_full_run/autofix-azure-pipelines-overwrite-autodetected-variables/complete.json]: yep, done

File: src/osemgrep/cli_ci/Ci_subcommand.ml (hunk 10 above)
Comment (jbergler) [reply src/osemgrep/cli_ci/Ci_subcommand.ml]: Done

File: cli/src/semgrep/app/scans.py (hunk 5 above)
Comment (jbergler) [reply cli/src/semgrep/app/scans.py:58]: agreed, see the earlier thread for more details on timing.

**Comment by jbergler** [comment]
The failure to build semgrep-pro exists on develop currently from a previous change, so I'm going to proceed with merging this.
